display_list/dl_op_spy.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/dl_op_spy.h"
+
+namespace flutter {
+
+bool DlOpSpy::DidDraw() {
+  return did_draw_;
+}
+
+void DlOpSpy::setColor(DlColor color) {
+  if (color.isTransparent()) {
+    dirty_ = false;
+  } else {
+    dirty_ = true;
+  }
+}
+void DlOpSpy::setColorSource(const DlColorSource* source) {
+  const DlColorColorSource* color_source = source->asColor();
+  if (color_source && color_source->color().isTransparent()) {
+    dirty_ = false;
+    return;
+  }
+  dirty_ = true;
+}
+void DlOpSpy::drawColor(DlColor color, DlBlendMode mode) {
+  if (did_draw_ || mode == DlBlendMode::kDst) {
+    return;
+  }
+  did_draw_ = !color.isTransparent() && mode != DlBlendMode::kClear;
+}
+void DlOpSpy::drawPaint() {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_;
+}
+void DlOpSpy::drawLine(const SkPoint& p0, const SkPoint& p1) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && p0 != p1;
+}
+void DlOpSpy::drawRect(const SkRect& rect) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !rect.isEmpty();
+}
+void DlOpSpy::drawOval(const SkRect& bounds) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !bounds.isEmpty();
+}
+void DlOpSpy::drawCircle(const SkPoint& center, SkScalar radius) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && radius > 0;
+}
+void DlOpSpy::drawRRect(const SkRRect& rrect) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !rrect.isEmpty();
+}
+void DlOpSpy::drawDRRect(const SkRRect& outer, const SkRRect& inner) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !outer.isEmpty() && outer != inner;
+}
+void DlOpSpy::drawPath(const SkPath& path) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !path.isEmpty();
```

Comment:
Double check if path.isEmpty really means there is no geometry. If it means the bounds are empty, then we might end up stroking something in stroke mode.